Blog: Integrating Siemens S7 PLCs with Node-RED: A Complete Guide (#2914)
FlowFuse/website: src/blog/2025/01/integrating-siemens-s7-plcs-with-node-red-guide.md

## Prequsite

Before integrating your Siemens S7 PLC with Node-RED, make sure you have the following :

1. Make sure before downloading the ladder program and all configurations and settings to plc have the following settings:

- Allow PUT/GET Communication from remote partners.

![PUT/GET Communication from remote partners is Allowed](./images/allow-put-get-communication.png){data-zoomable}
_PUT/GET Communication from remote partners is Allowed_

- Provide full access to the PLC (no protection), allowing unrestricted access to data exchange.

![Providing complete access to the PLC](./images/providing-full-access-to-plc.png){data-zoomable}
_Providing complete access to the PLC_

1. Make sure the appropriate ladder program (or any other logic) is written for your requirements and successfully downloaded to the PLC. This program should contain the logic for interacting with the PLC's I/O and sensors/actuators as needed.

> 2. Install Node-RED on your device (e.g., Revolutionary Pi) to connect and transform data across systems. Use [FlowFuse Device Agent](/product/device-agent/) to manage Node-RED remotely, allowing control, monitoring, and flow creation without on-site visits. FlowFuse also supports enterprise features like collaboration, device management, and DevOps pipelines. For setup, refer to the [installation guide](/docs/device-agent/install/).

Reviewer: Is this guide accurate? That is, is the S7 just a Debian based Linux box?

Author: Siemens PLCs, like most other industrial controllers, aren't regular computers and don't allow you to install software or applications. This point actually applies to devices with computing capabilities that need to communicate with the S7 PLC.

Author: Done, I have updated that point with a detailed explanation.

6. Once your flow is set up and the s7-out node for each variable is configured, click Deploy in the top-right corner to activate the flow.

<iframe width="100%" height="315" src="https://www.youtube.com/embed/AilWMNPzP1Q" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

In the video above, the dashboard interface is built to control the stack light. At the end of this article, I will provide the complete flow for you to download.

> If you're building a dashboard, keep in mind that when running Node-RED with the FlowFuse Device Agent, you can create the dashboard on the remote instance but you won’t be able to access it remotely across the device editor tunnel. You can of course access it locally on the device or on the local LAN.  For this demonstration, I wish to access the dashboard remotely across the internet and so I will create the dashboard in a hosted instance of Node-RED and using the FlowFuse Projects nodes to simplify passing values between the hosted and remote Node-RED instances. For more details on how to set this up, check out our article: [Exploring FlowFuse Project Nodes](/blog/2024/10/exploring-flowfuse-project-nodes/).

Reviewer 2: Sorry, i edited this as you merged.

I _think_ this reads better:

```suggestion
If you're building a dashboard, keep in mind that while you can create it on Node-RED within the remote instance on FlowFuse Device Agent, you won’t be able to access it remotely across the editor tunnel. You can of course access it locally on the device or on the local LAN. For this demonstration, I wish to access the dashboard remotely across the internet and so I will create the dashboard in a hosted instance of Node-RED and use the FlowFuse Projects nodes to simply and securely pass the necessary values to and from the remote Node-RED instance. For more details on how to set this up, check out our article: [Exploring FlowFuse Project Nodes]
```

{% endrenderFlow %}

## Troubleshooting 

When you try to establish a connection with the PLC, you may encounter the following error. This error occurs because your device has established the connection but is unable to communicate. To resolve this issue, ensure that you have configured all the settings mentioned in the prerequisites. If the problem persists, it could be because your PLC and the device running Node-RED are on different networks.

!["Error: This service is not implemented on the modeul or frame error was reported"](./images/error.png){data-zoomable}
_"Error: This service is not implemented on the modeul or frame error was reported"_

Make sure the IP addresses of your device and PLC are in the same subnet. If the PLC is connected to the internet via a router, all devices (PLC, Node-RED device, and router) should have IP addresses within the same subnet. For example, if your PLC has the address 192.168.1.1, ensure that the other devices have IP addresses in the range 192.168.1.x.

## Conclusion 

> Integrating Siemens S7 PLCs with Node-RED opens up powerful automation possibilities with minimal complexity. By following the steps outlined in this guide, you can easily connect your PLC to Node-RED, control devices, and visualize real-time data on dashboards. Whether you're writing data to control outputs or reading sensor values, Node-RED offers a flexible, user-friendly platform for industrial automation.

Author: ```suggestion
Integrating Siemens S7 PLCs with Node-RED opens up powerful automation possibilities with minimal complexity. By following the steps outlined in this guide, you can easily connect your PLC to Node-RED, control devices, and visualize real-time data on dashboards. Whether you're writing data to control outputs or reading sensor values, Node-RED offers a flexible, user-friendly platform for industrial automation.

Ready to take your automation projects to the next level? [Talk to our expert](/book-demo/) or [book a demo](/book-demo/ ) today to see how FlowFuse can help you scale, manage, and optimize your industrial automation.
```

Reviewer 2: > [Talk to our expert](/book-demo/) or [book a demo](/book-demo/ ) today

these are both the same link is that intended?

Also, the 2nd one has an extra space before the closing bracket!

Author: Yes, I want both of the words in the CTA. I have removed that extra space.

_Untick 'Optimized Block Access.'_

3. Install Node-RED on the device that will communicate with the S7 PLC. You cannot install Node-RED directly on the S7 PLC, as PLCs are typically controllers, not computers. For example, you can use a device like the Revolutionary Pi to connect and transfer data across systems. Use the [FlowFuse Device Agent](/product/device-agent/) to install Node-RED on your device. 

- Why FlowFuse Device Agent? It allows you to manage Node-RED remotely, enabling control, monitoring, and flow creation without the need for on-site visits. FlowFuse also offers a suite of enterprise-grade features such as collaboration, device management, and DevOps pipelines, which are essential in industrial environments. These features help streamline operations and ensure scalability in complex automation systems. For setup, refer to the [installation guide](/docs/device-agent/install/).

4. Verify that the device running Node-RED is in the same network as the PLC and can successfully ping the PLC. Also, a firewall should not block the S7 port (typically port 102).

```bash
ping <PLC_IP>
```

```bash
netstat -an | findstr 102

> ```

Reviewer 2: should we make a nod to the fact this that PING can be done inside node-red using `node-red-node-ping`?

also, the exec node can do these commands

also, I `findstr` is windows only. on linux it would be `grep` 

rather than provide all the different ways for the different OS's, perhaps simplify to ping only and mention firewalls as a side note?

My program in TIA Portal is structured as shown below, utilizing DB (Data Blocks) and Q (physical outputs) to control devices. However, Node-RED can retrieve almost all types of data from the PLC. The process is similar for most data types.

![Ladder Logic to Control Outputs for Managing Lights](./images/ladder-to-control-lights.png)_Ladder Logic to Control Outputs for Managing Lights_

Let’s break down what’s happening in the ladder logic above. First, we have open contacts, each with address variables defined in a separate Data Block. There are three open branches, each starting with an open contact. Each contact is connected to an output that alters the status of a Q physical address. Each Q corresponds to a physical output on the PLC, which is wired to the lights. When we change the status of a contact to "true," it activates the corresponding light by altering the state of the Q output, which reflects the change in the physical output.

### Installing the S7 Node

> To communicate from Node-RED to the PLC, we need to install the S7 node, which allows Node-RED to interface with Siemens S7 PLCs. This node provides the necessary functionality to read and write data to the PLC. While there are many S7 nodes available in the Node-RED catalog, in this article, we will be using `node-red-contrib-s7`, which is quite popular. However, it is important to note that this node is not actively updated, so you may choose other nodes that are more frequently maintained and updated. You can find these alternatives here: [Node-RED Siemens Nodes](https://flows.nodered.org/search?term=siemens&type=node).

Reviewer 2: ```suggestion
To communicate from Node-RED to the PLC, we need to install the S7 node, which allows Node-RED to interface with Siemens S7 PLCs. In this article, we will be using `node-red-contrib-s7`, which is quite popular. If this particular node is not suitable for your workflow you can find alternatives in the [Node-RED catalog](https://flows.nodered.org/search?term=siemens&type=node).
```

6. Once your flow is set up and the s7-out node for each variable is configured, click Deploy in the top-right corner to activate the flow.

<iframe width="100%" height="315" src="https://www.youtube.com/embed/AilWMNPzP1Q" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

In the video above, the dashboard interface is built to control the stack light. At the end of this article, I will provide the complete flow for you to download.

> If you're building a dashboard, keep in mind that when running Node-RED with the Flowfuse device agent, you won’t be able to build the dashboard within the remote instance. Instead, you'll need to create a separate hosted instance on Flowfuse to build your dashboard which is accessible for you. To enable communication between your Remote instance and the hosted dashboard, you can use the Flowfuse Projects nodes, which makes communication between diffrent Node-RED instances easy and simple. For more details on how to set this up, check out our article: [Exploring Flowfuse Project Nodes](/blog/2024/10/exploring-flowfuse-project-nodes/).

Reviewer 2: This is not 100% accurate. You can build your dashboard on a device and you can access it locally (just not across the device editor tunnel) 

Please consider this and add a suggestion below in a reply comment (<kbd>CTRL+G</kbd>) so I can review and merge it.

Author: ```suggestion
If you're building a dashboard, keep in mind that when running Node-RED with the FlowFuse Device Agent, you can create the dashboard within the remote instance. However, you won’t be able to access it across the device editor tunnel, though you can still access it locally on the device itself. So, I recommend building the dashboard in a separate FlowFuse hosted instance, which will allow you to access the dashboard remotely if needed. To enable communication between your remote instance and the hosted dashboard, you can use the FlowFuse Projects nodes, which make it easy to communicate between different Node-RED instances. For more details on how to set this up, check out our article: [Exploring FlowFuse Project Nodes](/blog/2024/10/exploring-flowfuse-project-nodes/).
```

Reviewer 2: Counter suggestion:
(includes LAN access and frame the narrative differently i.e.: "in this instance, I wish to access the dashboard remotely..."

```suggestion
If you're building a dashboard, keep in mind that when running Node-RED with the FlowFuse Device Agent, you can create the dashboard on the remote instance but you won’t be able to access it remotely across the device editor tunnel. You can of course access it locally on the device or on the local LAN.  For this demonstration, I wish to access the dashboard remotely across the internet and so I will create the dashboard in a hosted instance of Node-RED and using the FlowFuse Projects nodes to simplify passing values between the hosted and remote Node-RED instances. For more details on how to set this up, check out our article: [Exploring FlowFuse Project Nodes](/blog/2024/10/exploring-flowfuse-project-nodes/).
```

WDYT?

To address this issue, you can optimize data retrieval by storing output status values in a single word or double word within the PLC. For our example, I have created a custom function in my program that assigns the output values to individual bits of the word. 

![Ladder diagram showing a custom function that stores the status of outputs in a single word within the PLC.](./images/custom-function-storing-bits-in-word.png){data-zoomable}
_Custom ladder diagram function storing output statuses in a single word for optimized data retrieval._

There are several ways to implement this, and depending on your system’s needs, some methods may be more efficient than others. In this case, the output values are stored in a single word within the PLC, as shown in the ladder diagram above. This is not the only correct method—it's simply one approach that works for this particular scenario. Feel free to adapt or explore other methods that might better suit your setup. 

> Additionally, if the data you’re reading is mission-critical and you can't afford to lose any, consider using stack logic in your PLC program. This method ensures that even if there’s a network disruption or communication failure, no data is lost. Stack logic temporarily stores the data, and once the network connection is restored, the stored data is transmitted, ensuring no gaps or interruptions in your system. This guarantees data integrity and prevents costly data loss, even in less-than-ideal conditions.

Reviewer 2: ```suggestion
Additionally, if the data you’re reading is mission-critical and you can't afford to lose any, consider using a FIFO stack or buffer in your PLC program. This method ensures that even if there is a network outage or computer problem, no data is lost as it will remain siting in the stack until your Node-RED is back on line and retrieves it.  This ensures no gaps or interruptions in your data and guarantees data integrity.
```
